ports/esp32/machine_ulp.c

```diff
@@ -0,0 +1,100 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+#include "machine_ulp.h"
+
+#include "py/obj.h"
+#include "py/runtime.h"
+#include "py/mperrno.h"
+
+#include "esp32/ulp.h"
+#include "esp_err.h"
+
+typedef struct _machine_ulp_obj_t {
+    mp_obj_base_t base;
+} machine_ulp_obj_t;
+
+const mp_obj_type_t machine_ulp_type;
+
+// singleton ULP object
+STATIC const machine_ulp_obj_t machine_ulp_obj = {{&machine_ulp_type}};
+
+STATIC mp_obj_t machine_ulp_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *args) {
+    // check arguments
+    mp_arg_check_num(n_args, n_kw, 0, 0, false);
+
+    // return constant object
+    return (mp_obj_t)&machine_ulp_obj;
+}
+
+STATIC mp_obj_t machine_ulp_set_wakeup_period(mp_obj_t self_in, mp_obj_t period_index_in, mp_obj_t period_us_in) {
+    mp_uint_t period_index = mp_obj_get_int(period_index_in);
+    mp_uint_t period_us = mp_obj_get_int(period_us_in);
+    int _errno = ulp_set_wakeup_period(period_index, period_us);
+    if (_errno != ESP_OK) {
+	    mp_raise_OSError(_errno);
+    }
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_3(machine_ulp_set_wakeup_period_obj, machine_ulp_set_wakeup_period);
+
+STATIC mp_obj_t machine_ulp_load_binary(mp_obj_t self_in, mp_obj_t load_addr_in, mp_obj_t program_binary_in) {
+    mp_uint_t load_addr = mp_obj_get_int(load_addr_in);
+
+    mp_buffer_info_t bufinfo;
+    mp_get_buffer_raise(program_binary_in, &bufinfo, MP_BUFFER_READ);
+
+    int _errno = ulp_load_binary(load_addr, bufinfo.buf, bufinfo.len/sizeof(uint32_t));
+    if (_errno != ESP_OK) {
+	    mp_raise_OSError(_errno);
+    }
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_3(machine_ulp_load_binary_obj, machine_ulp_load_binary);
+
+STATIC mp_obj_t machine_ulp_run(mp_obj_t self_in, mp_obj_t entry_point_in) {
+    mp_uint_t entry_point = mp_obj_get_int(entry_point_in);
+    int _errno = ulp_run(entry_point/sizeof(uint32_t));
+    if (_errno != ESP_OK) {
+	    mp_raise_OSError(_errno);
+    }
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_2(machine_ulp_run_obj, machine_ulp_run);
+
+STATIC const mp_map_elem_t machine_ulp_locals_dict_table[] = {
+    { MP_OBJ_NEW_QSTR(MP_QSTR_set_wakeup_period), (mp_obj_t)&machine_ulp_set_wakeup_period_obj },
+    { MP_OBJ_NEW_QSTR(MP_QSTR_load_binary), (mp_obj_t)&machine_ulp_load_binary_obj },
+    { MP_OBJ_NEW_QSTR(MP_QSTR_run), (mp_obj_t)&machine_ulp_run_obj },
+    { MP_OBJ_NEW_QSTR(MP_QSTR_COPROC_RESERVE_MEM), MP_ROM_INT(CONFIG_ULP_COPROC_RESERVE_MEM) },
```

Comment:
I guess this constant is needed/useful a script can check how much room is left to load a ULP binary, right?

Since the constant lives in the ULP class already it could be simplified to `RESERVE_MEM`.

Reply:
You are correct. I wanted to keep the same name that is used in the esp-idf but if you prefere a shorter name I can for sure change it.

Reply:
COPROC_RESERVE_MEM is already a shortened version of the name so I think it's fair to shorten it further to RESERVE_MEM.  It doesn't lose any of it's meaning and shorter names are easier to type, take less space in flash and in scripts, etc.